File: tools/theme-contrast-natsunohi/measure.py
#!/usr/bin/env python3
"""Kontrast-Harness für das Theme „natsunohi" (夏の日).

Warum es diese Datei gibt
─────────────────────────
Im Repo existierte am 18.08. KEIN Kontrast-Mess-Harness (verifiziert). Die
Asagiri-Zahlen im Datei-Kopf jenes Themes sind mit einem Ad-hoc-Messfühler
entstanden, und die Lehre desselben Tages steht dort ausdrücklich: ein Fühler
OHNE den kompletten CSS-Import hat gelogen. Darum lädt dieses Harness immer
index.css (Basistoken) UND die Theme-Datei, und es misst auf echten
headless-Chrome-Pixeln statt auf gerechneten Farben.

Zwei Messungen, weil beide etwas anderes beweisen
─────────────────────────────────────────────────
1) PALETTE — eine Seite mit Farbkacheln. Jede Kachel wird gerendert und ihr
   Mittelpixel aus dem Screenshot gelesen. Das gibt die EXAKTEN sRGB-Werte, die
   der Browser aus den OKLCH-Angaben macht (inkl. Gamut-Mapping), und damit
   auch die *-soft-Füllungen als echte Komposition über --bg-surface.
2) SZENE — eine leere Seite mit Atmosphäre + Szene + Schleier + Deko-Ebenen.
   Screenshot bei 1440/1280/1024 px, dann der SCHLECHTESTE Pixel innerhalb der
   920-px-App-Spalte gegen --text-4 und --text-1.

   Die Szene wird ZWEIMAL gemessen:
     • „floor"  — die aufhellenden Deko-Ebenen (Flirren, Zikaden-Licht) sind
       abgeschaltet. Das ist der Boden: beide Ebenen können auf hellem Grund
       nur AUFHELLEN, also den Kontrast gegen dunkle Schrift nur verbessern.
       Kein Animationsframe kann je unter diesen Wert fallen — deshalb ist das
       der Wert, der im Datei-Kopf steht.
     • „full"   — alles an, Animationen per prefers-reduced-motion eingefroren.
       Nur zur Kontrolle, dass die Ebenen wirklich aufhellen.

Aufruf:  python3 measure.py            (misst alles)
         python3 measure.py --keep     (HTML/PNG im Ausgabeordner behalten)
"""

from __future__ import annotations

import json
import os
import shutil
import signal
import struct
import subprocess
import sys
import tempfile
import time
import zlib

HERE = os.path.dirname(os.path.abspath(__file__))
REPO = os.path.abspath(os.path.join(HERE, "..", ".."))
THEMES = os.path.join(REPO, "frontend", "public", "themes")
INDEX_CSS = os.path.join(REPO, "frontend", "src", "index.css")
# Überschreibbar, damit das Harness gegen ein FREMDES, bereits veröffentlichtes
# Theme geeicht werden kann: `THEME_ID=asagiri python3 measure.py` muss die im
# Asagiri-Kopf dokumentierten Werte (4,81 / 12,73) reproduzieren. Tut es das
# nicht, lügt der Messfühler — und nicht das Theme.
THEME_ID = os.environ.get("THEME_ID", "natsunohi")
THEME_CSS = os.path.join(THEMES, THEME_ID + ".css")
CHROME = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"

# Die App-Lesespalte (index.css: `.app { max-width: 920px; margin: 0 auto }`).
APP_COLUMN = 920
WIDTHS = [(1440, 900), (1280, 800), (1024, 768)]


# ── PNG ohne Pillow ─────────────────────────────────────────────────────────
def read_png(path: str):
    """Minimaler PNG-Leser (8 bit, Farbtyp 2/6, nicht interlaced) → (w, h, rows).

    Chrome schreibt genau dieses Format. Rückgabe: rows[y] = bytes in RGB(A).
    """
    with open(path, "rb") as fh:
        data = fh.read()
    assert data[:8] == b"\x89PNG\r\n\x1a\n", "kein PNG"
    pos = 8
    idat = bytearray()
    width = height = depth = ctype = None
    while pos < len(data):
        (length,) = struct.unpack(">I", data[pos : pos + 4])
        ctag = data[pos + 4 : pos + 8]
        body = data[pos + 8 : pos + 8 + length]
        pos += 12 + length
        if ctag == b"IHDR":
            width, height, depth, ctype, _, _, interlace = struct.unpack(">IIBBBBB", body)
            assert depth == 8 and ctype in (2, 6) and interlace == 0, (depth, ctype, interlace)
        elif ctag == b"IDAT":
            idat += body
        elif ctag == b"IEND":
            break
    raw = zlib.decompress(bytes(idat))
    channels = 3 if ctype == 2 else 4
    stride = width * channels
    rows: list[bytearray] = []
    prev = bytearray(stride)
    off = 0
    for _ in range(height):
        ftype = raw[off]
        off += 1
        line = bytearray(raw[off : off + stride])
        off += stride
        if ftype == 1:  # Sub
            for i in range(channels, stride):
                line[i] = (line[i] + line[i - channels]) & 0xFF
        elif ftype == 2:  # Up
            for i in range(stride):
                line[i] = (line[i] + prev[i]) & 0xFF
        elif ftype == 3:  # Average
            for i in range(stride):
                left = line[i - channels] if i >= channels else 0
                line[i] = (line[i] + ((left + prev[i]) >> 1)) & 0xFF
        elif ftype == 4:  # Paeth
            for i in range(stride):
                a = line[i - channels] if i >= channels else 0
                b = prev[i]
                c = prev[i - channels] if i >= channels else 0
                p = a + b - c
                pa, pb, pc = abs(p - a), abs(p - b), abs(p - c)
                pr = a if (pa <= pb and pa <= pc) else (b if pb <= pc else c)
                line[i] = (line[i] + pr) & 0xFF
        rows.append(line)
        prev = line
    return width, height, rows, channels


# ── WCAG 2.1 ────────────────────────────────────────────────────────────────
def _lin(c: int) -> float:
    s = c / 255.0
    return s / 12.92 if s <= 0.04045 else ((s + 0.055) / 1.055) ** 2.4


def luminance(rgb) -> float:
    r, g, b = rgb[0], rgb[1], rgb[2]
    return 0.2126 * _lin(r) + 0.7152 * _lin(g) + 0.0722 * _lin(b)


def contrast(a, b) -> float:
    la, lb = luminance(a), luminance(b)
    hi, lo = max(la, lb), min(la, lb)
    return (hi + 0.05) / (lo + 0.05)


def hexof(rgb) -> str:
    return "#%02x%02x%02x" % (rgb[0], rgb[1], rgb[2])


# ── Chrome ──────────────────────────────────────────────────────────────────
def shoot(html_path: str, png_path: str, size, profile_dir: str) -> None:
    """Ein Screenshot, ein eigener Chrome.

    Chrome 151 schreibt die PNG-Datei zuverlässig, BEENDET sich danach aber
    nicht mehr (der alte `--screenshot`-und-raus-Weg ist tot). Darum: starten,
    auf die fertige Datei warten (Größe zwei Runden stabil), dann NUR den
    eigenen Prozessbaum beenden. Fremde Chrome-Fenster bleiben unangetastet —
    wir kennen unsere PID und benutzen ein eigenes --user-data-dir.
    """
    w, h = size
    if os.path.exists(png_path):
        os.remove(png_path)
    cmd = [
        CHROME,
        "--headless=new",
        "--disable-gpu",
        "--hide-scrollbars",
        "--force-color-profile=srgb",
        "--force-device-scale-factor=1",
        "--allow-file-access-from-files",
        "--no-first-run",
        "--no-default-browser-check",
        "--disable-extensions",
        "--user-data-dir=" + profile_dir,  # eigener Profilordner: fremde Chromes bleiben unberührt
        "--window-size=%d,%d" % (w, h),
        "--virtual-time-budget=2500",
        "--screenshot=" + png_path,
        "file://" + html_path,
    ]
    proc = subprocess.Popen(
        cmd, stdout=subprocess.DEVNULL, stderr=subprocess.PIPE, start_new_session=True
    )
    last = -1
    stable = 0
    try:
        for _ in range(600):  # bis 60 s
            time.sleep(0.1)
            if proc.poll() is not None and os.path.exists(png_path):
                break
            if os.path.exists(png_path):
                size_now = os.path.getsize(png_path)
                if size_now > 0 and size_now == last:
                    stable += 1
                    if stable >= 3:
                        break
                else:
                    stable = 0
                last = size_now
    finally:
        if proc.poll() is None:
            try:
                os.killpg(os.getpgid(proc.pid), signal.SIGKILL)
            except (ProcessLookupError, PermissionError):
                proc.kill()
            proc.wait(timeout=20)
    if not os.path.exists(png_path) or os.path.getsize(png_path) == 0:
        err = proc.stderr.read().decode("utf-8", "replace") if proc.stderr else ""
        sys.exit("Chrome hat nichts gerendert:\n" + err[-3000:])


# ── Seiten ──────────────────────────────────────────────────────────────────
HEAD = """<!doctype html><html data-theme='{tid}'><head><meta charset='utf-8'>
<link rel='stylesheet' href='file://{index}'>
<link rel='stylesheet' href='file://{theme}'>
<style>{extra}</style></head><body>{body}</body></html>"""

# Die Kacheln: (Bezeichner, Aufbau). "solid" = eine Fläche, "over" = Füllung
# über --bg-surface (so bauen die Pills/Badges in index.css).
SOLIDS = [
    "bg-base",
    "bg-surface",
    "bg-elevated",
    "bg-subtle",
    "bg-user",
    "bg-hairline",
    "bg-hairline-dashed",
    "text-1",
    "text-2",
    "text-3",
    "text-4",
    "accent",
    "accent-ink",
    "success",
    "warn",
    "error",
]
OVERS = ["accent-soft", "accent-faint", "success-soft", "warn-soft", "error-soft", "accent-glow"]

CELL = 40


def palette_html(extra: str = "") -> str:
    cells = []
    for name in SOLIDS:
        cells.append("<i style='background:var(--%s)'></i>" % name)
    for name in OVERS:
        cells.append(
            "<i style='background:var(--bg-surface)'>"
            "<b style='background:var(--%s)'></b></i>" % name
        )
    css = (
        "html,body{margin:0;padding:0;background:#808080}"
        "i{display:block;float:left;width:%dpx;height:%dpx}"
        "b{display:block;width:100%%;height:100%%}" % (CELL, CELL)
    )
    return HEAD.format(
        tid=THEME_ID,
        index=INDEX_CSS,
        theme=THEME_CSS,
        extra=css + extra,
        body="".join(cells),
    )


# Die Szene: leere App-Spalte, sonst nichts. Zusätzlich das reduced-motion-Bild
# erzwingen wäre unnötig — Chrome headless animiert bei virtual time ohnehin
# deterministisch; die „floor"-Variante schaltet die aufhellenden Ebenen ganz ab.
SCENE_BODY = "<div class='app'></div>"
# ACHTUNG, hier lag ein stiller Messfehler: solange der Schleier auf demselben
# Element saß wie die Zeichnung, blendete diese Variante mit `:root::after` die
# DEKO aus. Seit der zweiten Fassung ist `:root::after` DER SCHLEIER — dieselbe
# Regel hätte ihn mitabgeschaltet und einen Kontrast von 1,0 gemeldet, also ein
# Theme für kaputt erklärt, das in Ordnung ist. Der Boden ist die Seite MIT
# Schleier und OHNE Luft: die Luft-Ebene ist nahezu weiß auf hellem Grund und
# kann nur aufhellen, der schlechteste Fall ist also die Seite ohne sie.
FLOOR_EXTRA = ":root[data-theme='%s'] body::before{display:none!important}" % THEME_ID


# Beide Szenen-Ebenen wogen, der Schleier steht (er ist ein eigenes Element und
# wird nicht mitgeschert). Trotzdem ist der schlechteste Moment nicht die Ruhe,
# sondern der volle Ausschlag: dort steht ein anderer Bildpunkt unter der
# Spaltenkante. Diese Variante hält beide Animationen an und setzt die
# Extremwerte fest ein — hintere Ebene ±1,1°, vordere ±3,2°, wie in der CSS.
SKEW_EXTRA = FLOOR_EXTRA + (
    ":root[data-theme='%s'] body::after{animation:none!important;"
    "transform:skewX(%s)!important}"
    ":root[data-theme='%s']::before{animation:none!important;"
    "transform:skewX(%s)!important}" % (THEME_ID, "{DEG}", THEME_ID, "{DEG3}")
)


def scene_html(variant: str) -> str:
    if variant == "floor":
        extra = FLOOR_EXTRA
    elif variant.startswith("skew"):
        sign = 1.0 if variant == "skew+" else -1.0
        extra = SKEW_EXTRA.replace("{DEG}", f"{1.1 * sign}deg").replace(
            "{DEG3}", f"{3.2 * sign}deg"
        )
    else:
        extra = ""
    return HEAD.format(
        tid=THEME_ID, index=INDEX_CSS, theme=THEME_CSS, extra=extra, body=SCENE_BODY
    )


# ── Messung ─────────────────────────────────────────────────────────────────
def measure() -> dict:
    out = tempfile.mkdtemp(prefix="natsunohi-contrast-")
    profile = os.path.join(out, "chrome-profile")  # NUR unser eigenes Profil
    result: dict = {"palette": {}, "scene": {}}

    # 1) Palette
    p_html = os.path.join(out, "palette.html")
    with open(p_html, "w") as fh:
        fh.write(palette_html())
    p_png = os.path.join(out, "palette.png")
    shoot(p_html, p_png, (CELL * len(SOLIDS + OVERS) + 80, 200), profile)
    w, h, rows, ch = read_png(p_png)
    colors: dict[str, tuple] = {}
    for i, name in enumerate(SOLIDS + OVERS):
        x = i * CELL + CELL // 2
        y = CELL // 2
        px = rows[y][x * ch : x * ch + 3]
        colors[name] = (px[0], px[1], px[2])
    result["colors"] = {k: hexof(v) for k, v in colors.items()}

    # Token-Matrix
    grounds = ["bg-base", "bg-surface", "bg-elevated", "bg-subtle", "bg-user"]
    fores = ["text-1", "text-2", "text-3", "text-4", "accent", "success", "warn", "error"]
    matrix = {}
    for f in fores:
        row = {g: round(contrast(colors[f], colors[g]), 2) for g in grounds}
        row["WORST"] = round(min(row.values()), 2)
        matrix[f] = row
    result["matrix"] = matrix
    result["on_own_soft"] = {
        "accent": round(contrast(colors["accent"], colors["accent-soft"]), 2),
        "success": round(contrast(colors["success"], colors["success-soft"]), 2),
        "warn": round(contrast(colors["warn"], colors["warn-soft"]), 2),
        "error": round(contrast(colors["error"], colors["error-soft"]), 2),
    }
    result["accent_ink_on_accent"] = round(contrast(colors["accent-ink"], colors["accent"]), 2)

    # 2) Szene
    for variant in ("floor", "skew+", "skew-", "full"):
        s_html = os.path.join(out, "scene-%s.html" % variant.replace("+", "p").replace("-", "m"))
        with open(s_html, "w") as fh:
            fh.write(scene_html(variant))
        per_width = {}
        for (vw, vh) in WIDTHS:
            png = os.path.join(out, "scene-%s-%d.png" % (variant.replace("+","p").replace("-","m"), vw))
            shoot(s_html, png, (vw, vh), profile)
            w, h, rows, ch = read_png(png)
            left = max(0, (w - APP_COLUMN) // 2)
            right = w - left
            worst4 = (99.0, None)
            worst1 = (99.0, None)
            seen: dict[tuple, bool] = {}
            for y in range(0, h, 2):  # jede 2. Zeile — 0,5 Mio Pixel reichen für die Suche
                row = rows[y]
                for x in range(left, right, 2):
                    px = (row[x * ch], row[x * ch + 1], row[x * ch + 2])
                    if px in seen:
                        continue
                    seen[px] = True
                    c4 = contrast(colors["text-4"], px)
                    if c4 < worst4[0]:
                        worst4 = (c4, px, x, y)
                    c1 = contrast(colors["text-1"], px)
                    if c1 < worst1[0]:
                        worst1 = (c1, px, x, y)
            per_width[vw] = {
                "text-4": round(worst4[0], 2),
                "text-4_pixel": hexof(worst4[1]),
                "text-4_at": [worst4[2], worst4[3]],
                "text-1": round(worst1[0], 2),
                "text-1_pixel": hexof(worst1[1]),
                "column": [left, right],
                "distinct_colors": len(seen),
            }
        result["scene"][variant] = per_width

    result["_outdir"] = out
    return result


if __name__ == "__main__":
    r = measure()
    print(json.dumps({k: v for k, v in r.items() if k != "_outdir"}, indent=2, ensure_ascii=False))
    if "--keep" in sys.argv:
        print("\nArtefakte: " + r["_outdir"], file=sys.stderr)
    else:
        shutil.rmtree(r["_outdir"], ignore_errors=True)

File: tools/theme-contrast/measure.mjs
/**
 * Kontrast-Harness: misst das ECHTE Render eines Themas, nicht ein Modell davon.
 * ═══════════════════════════════════════════════════════════════════════════
 * DIE LEHRE, DIE ES GEBAUT HAT (18.08.): ein Messfühler, der die Themen-CSS
 * ohne die Basis-Token aus `index.css` lädt, misst einen Hintergrund, den es so
 * nie gibt — er hat gelogen und der Fehler fiel erst am echten Bild auf. Darum
 * lädt dieses Harness IMMER den vollständigen Stapel in genau der Reihenfolge
 * der App:  index.css (Basis/Yoru) → styles/themes.css (Panel + :not-Basis) →
 * public/themes/<id>.css (das Thema selbst, aus SEINEM echten Verzeichnis,
 * damit `url('…-szene.svg')` relativ auflöst wie im Browser).
 *
 * WAS GEMESSEN WIRD. Nicht „Token gegen Token" (das rechnet color.mjs), sondern
 * der schlechteste PIXEL: für jede Textstufe wird über alle Pixel der 920-px-
 * App-Spalte das Minimum des WCAG-Kontrasts gesucht. Das ist die einzige Zahl,
 * die eine Zusage trägt — sie schließt Atmosphäre, Szene, Schleier und jede
 * animierte Schicht mit ein, egal wie sie zustande kommen.
 *
 * ZEIT IST EIN MESSPUNKT. Ein Thema mit seltenen Ereignissen (Nagareboshi:
 * Meteore) hat seinen schlechtesten Moment nicht bei t=0. Szenarien frieren
 * darum einzelne Schichten per `animation-play-state: paused` + negativem
 * `animation-delay` auf einem exakten Keyframe ein: gemessen wird echtes CSS zu
 * einer gewählten Sekunde, keine Nachbildung.
 *
 * ZWEI SCHRAUBEN JE SZENARIO, seit Nagareboshis zweiter Fassung (19.08.):
 *   `vt`     — das virtuelle Zeitbudget in ms (Vorgabe 3000). Nötig, weil eine
 *              Szene ihre Bewegung auch INNERHALB des SVG tragen kann; solche
 *              Uhren erreicht kein `animation-delay` von außen, wohl aber die
 *              Wahl der Sekunde, zu der Chrome knipst.
 *   `flags`  — zusätzliche Chrome-Schalter. Der wichtige ist
 *              `--force-prefers-reduced-motion`: er hält AUCH die Uhren im SVG
 *              an. Ist dessen Ruhezustand der hellste (Nagareboshi baut ihn so),
 *              misst dieses Szenario den echten Worst Case des Sternfelds
 *              statt einer zufällig erwischten Phase.
 *
 * Eigenes `--user-data-dir` unter os.tmpdir() (Pod-Regel: keine fremden
 * Chrome-Profile, keine fremden Prozesse anfassen).
 *
 * NUTZUNG
 *   node tools/theme-contrast/measure.mjs tools/theme-contrast/nagareboshi.scenarios.mjs
 *   node tools/theme-contrast/measure.mjs <config> --keep   (Screenshots behalten)
 */

import { createServer } from 'node:http';
import { spawn } from 'node:child_process';
import { readFileSync, statSync, mkdtempSync, rmSync, existsSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join, resolve, dirname, extname } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { decodePng, pixel } from './png.mjs';
import { contrast, hex, luminance } from './color.mjs';

const HERE = dirname(fileURLToPath(import.meta.url));
const REPO = resolve(HERE, '..', '..');
const CHROME = '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome';

/** Höhe des Token-Messstreifens; darunter beginnt das gemessene Feld. */
const PROBE_H = 24;
const PROBE_SW = 20;
/** Halbe Breite der App-Spalte (`.app { max-width: 920px; margin: auto }`). */
const HALF_COLUMN = 460;

const MIME = {
  '.css': 'text/css',
  '.svg': 'image/svg+xml',
  '.html': 'text/html',
  '.png': 'image/png',
  '.json': 'application/json',
};

/* ── Die Seite, die gemessen wird ────────────────────────────────────────── */

function harnessHtml(cfg, scenario) {
  const swatches = cfg.probeTokens
    .map((t) => `<i style="background:var(${t})"></i>`)
    .join('');
  return `<!doctype html>
<html lang="de" data-theme="${cfg.theme}">
<head>
<meta charset="utf-8">
<title>theme-contrast</title>
<link rel="stylesheet" href="/src/index.css">
<link rel="stylesheet" href="/src/styles/themes.css">
<link rel="stylesheet" href="/themes/${cfg.theme}.css">
<style>
  /* Der Messstreifen: jede Kachel ist EIN Token, deckend gemalt. So kommt der
     Wert, den Chrome wirklich aus oklch() macht, im selben Screenshot mit —
     keine zweite Farbmathematik, die von der des Browsers abweichen könnte. */
  #probe { position: fixed; top: 0; left: 0; z-index: 9999; display: flex; margin: 0; padding: 0; }
  #probe i { display: block; width: ${PROBE_SW}px; height: ${PROBE_H}px; }
  html, body { height: 100%; margin: 0; }
${scenario.css || ''}
</style>
</head>
<body>
<div class="app"></div>
<div id="probe">${swatches}</div>
</body>
</html>`;
}

/* ── Statischer Server über dem Worktree ─────────────────────────────────── */

function startServer(cfg) {
  const server = createServer((req, res) => {
    const url = new URL(req.url, 'http://127.0.0.1');
    const scenario = cfg.scenarios.find((s) => s.name === url.searchParams.get('s')) || { css: '' };

    if (url.pathname === '/harness.html') {
      res.writeHead(200, { 'content-type': 'text/html; charset=utf-8' });
      res.end(harnessHtml(cfg, scenario));
      return;
    }

    // /themes/* liegt im echten Auslieferungs-Verzeichnis, damit relative
    // url()-Verweise der Themen-CSS (die Szenen-SVG!) genau wie in Produktion
    // auflösen. /src/* ist die Basis-CSS.
    const map = [
      ['/themes/', join(REPO, 'frontend', 'public', 'themes')],
      ['/src/', join(REPO, 'frontend', 'src')],
    ];
    for (const [prefix, root] of map) {
      if (url.pathname.startsWith(prefix)) {
        const file = join(root, url.pathname.slice(prefix.length));
        if (!file.startsWith(root) || !existsSync(file)) break;
        res.writeHead(200, { 'content-type': MIME[extname(file)] || 'application/octet-stream' });
        res.end(readFileSync(file));
        return;
      }
    }
    res.writeHead(404).end('nope');
  });
  return new Promise((ok) => server.listen(0, '127.0.0.1', () => ok(server)));
}

/* ── Chrome ──────────────────────────────────────────────────────────────── */

const sleep = (ms) => new Promise((ok) => setTimeout(ok, ms));

/**
 * Ein Screenshot.
 *
 * WARUM NICHT `spawnSync`: Chrome schreibt die PNG nach ~1 s, beendet sich auf
 * diesem Mac aber erst ~25 s später (es hängt nach getaner Arbeit noch an
 * eigenen Aufräum-/Telemetriepfaden). Synchron gewartet kostet ein Messlauf
 * über drei Breiten und fünf Szenarien eine Viertelstunde statt einer Minute.
 * Deshalb: starten, auf die FERTIGE Datei warten (Größe muss zweimal in Folge
 * gleich sein — sonst liest man ein halb geschriebenes PNG) und dann den EIGENEN
 * Kindprozess beenden. Pod-Regel: nur der eigene Prozess, nie ein fremder
 * Chrome; die Prozess-Id kommt aus genau diesem `spawn`.
 */
async function shoot(url, width, height, out, profile, opts = {}) {
  const child = spawn(
    CHROME,
    [
      '--headless',
      '--no-sandbox', // sonst startet Chrome in dieser Umgebung nicht durch
      '--disable-gpu',
      '--hide-scrollbars',
      '--force-device-scale-factor=1',
      '--no-first-run',
      '--no-default-browser-check',
      `--user-data-dir=${profile}`,
      `--window-size=${width},${height}`,
      `--virtual-time-budget=${opts.vt ?? 3000}`,
      ...(opts.flags ?? []),
      `--screenshot=${out}`,
      url,
    ],
    { stdio: 'ignore' },
  );

  let lastSize = -1;
  for (let i = 0; i < 600; i++) {
    await sleep(100);
    if (existsSync(out)) {
      const size = statSync(out).size;
      if (size > 0 && size === lastSize) break;
      lastSize = size;
    }
    if (child.exitCode !== null && existsSync(out)) break;
  }
  child.kill('SIGKILL');
  if (!existsSync(out)) throw new Error(`Chrome lieferte kein Bild für ${url} @${width}`);
}

/* ── Auswertung ──────────────────────────────────────────────────────────── */

/**
 * Schlechtester Kontrast einer Textfarbe gegen ein Pixelfeld.
 * „Schlechtester" heißt: das Pixel, dessen Luminanz der Schrift am NÄCHSTEN
 * liegt — für helle Schrift auf dunklem Grund ist das automatisch die hellste
 * Stelle, für dunkle Schrift die dunkelste. Ein Polaritäts-Sonderfall ist damit
 * gar nicht erst nötig.
 */
function worstAgainst(img, box, textRgb) {
  const lt = luminance(textRgb);
  let worst = Infinity;
  let at = null;
  for (let y = box.y0; y < box.y1; y++) {
    for (let x = box.x0; x < box.x1; x++) {
      const p = pixel(img, x, y);
      const lp = luminance(p);
      const ratio = (Math.max(lt, lp) + 0.05) / (Math.min(lt, lp) + 0.05);
      if (ratio < worst) {
        worst = ratio;
        at = { x, y, rgb: p };
      }
    }
  }
  return { ratio: worst, at };
}

function fmt(n) {
  return n.toFixed(2).replace('.', ',');
}

/* ── Lauf ────────────────────────────────────────────────────────────────── */

const configPath = process.argv[2];
if (!configPath) {
  console.error('nutzung: node measure.mjs <scenarios.mjs> [--keep]');
  process.exit(2);
}
const keep = process.argv.includes('--keep');
const cfg = (await import(pathToFileURL(resolve(configPath)).href)).default;

const server = await startServer(cfg);
const port = server.address().port;
const work = mkdtempSync(join(tmpdir(), 'hoshi-theme-contrast-'));
const profile = join(work, 'chrome-profile');

let fails = 0;
try {
  for (const scenario of cfg.scenarios) {
    console.log(`\n━━ ${scenario.name} ━━ ${scenario.note || ''}`);
    for (const [width, height] of cfg.viewports) {
      const png = join(work, `${scenario.name}-${width}x${height}.png`);
      await shoot(
        `http://127.0.0.1:${port}/harness.html?s=${encodeURIComponent(scenario.name)}`,
        width,
        height,
        png,
        profile,
        { vt: scenario.vt, flags: scenario.flags },
      );
      const img = decodePng(readFileSync(png));

      // Token-Istwerte aus dem Messstreifen lesen.
      const tokens = {};
      cfg.probeTokens.forEach((name, i) => {
        tokens[name] = pixel(img, i * PROBE_SW + PROBE_SW / 2, PROBE_H / 2);
      });

      const cx = Math.round(img.width / 2);
      const column = {
        x0: Math.max(0, cx - HALF_COLUMN),
        x1: Math.min(img.width, cx + HALF_COLUMN),
        y0: PROBE_H + 16,
        y1: img.height,
      };

      const parts = [];
      for (const t of cfg.textTokens) {
        const { ratio, at } = worstAgainst(img, column, tokens[t]);
        const ok = ratio >= (cfg.floor ?? 4.5);
        if (!ok) fails++;
        parts.push(`${t.replace('--', '')} ${fmt(ratio)}:1${ok ? '' : ' ✗'} @${hex(at.rgb)}`);
      }

      // Randraum nur zur Information: dort steht das Bild in voller Stärke und
      // es sitzt nie eine Glyphe (die App-Spalte ist 920 px und zentriert).
      let margin = '—';
      if (column.x0 > 8) {
        const m = worstAgainst(img, { x0: 0, x1: column.x0, y0: column.y0, y1: column.y1 }, tokens[cfg.textTokens.at(-1)]);
        margin = `${fmt(m.ratio)}:1`;
      }
      console.log(`  ${String(width).padStart(4)}×${height}  ${parts.join(' · ')}   [Rand ${margin}]`);
      if (!keep) rmSync(png, { force: true });
    }
  }
} finally {
  server.close();
  if (!keep) rmSync(work, { recursive: true, force: true });
  else console.log(`\nScreenshots: ${work}`);
}

console.log(
  fails === 0
    ? `\n✓ alle Werte ≥ ${cfg.floor ?? 4.5}:1 in der 920-px-Spalte`
    : `\n✗ ${fails} Messung(en) unter ${cfg.floor ?? 4.5}:1`,
);
process.exit(fails === 0 ? 0 : 1);
